### ISSUE TITLE
background worker huge memory allocation

### ISSUE THREAD

**Issue by DanielCristianPinogy** [citusdata/pg_cron#26]
We're seeing this behavior which is more perceptible when using many jobs scheduled on pg_cron.

This is the situation after one week with 400 jobs:

```
  PID USER      PR  NI    VIRT    RES    SHR   USED S  %CPU %MEM     TIME+ COMMAND
 1786 postgres  20   0 9766.3m 4.181g   3228 4.181g S   0.0 22.0   0:38.14 postgres: bgworker: pg_cron_scheduler
```

As shown the shared memory is aroung 3.2MB and the process memory is 4.1GB.

The version where the issue was detected is PostgreSQL 9.5.7 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-11), 64-bit, but also occurs on PostgreSQL 9.5.7 on x86_64-pc-linux-gnu, compiled by gcc (Debian 4.9.2-10) 4.9.2, 64-bit.

Bellow there are the SQL commands creating jobs and the top measurements showing the memory increase:

```
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) values ('* * * * *', 'select 1', 'localhost', '5432', 'postgres', 'postgres');
              now              
-------------------------------
 2017-06-01 10:58:28.591112-03
(1 row)
INSERT 0 1
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
             now              
------------------------------
 2017-06-01 10:58:48.90418-03
(1 row)
INSERT 0 1
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 10:59:43.526128-03
(1 row)
INSERT 0 2
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 11:00:04.406233-03
(1 row)
INSERT 0 4
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 11:00:20.862736-03
(1 row)
INSERT 0 8
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 11:00:49.438208-03
(1 row)
INSERT 0 16
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 11:01:27.310265-03
(1 row)
INSERT 0 32
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 11:01:56.518833-03
(1 row)
INSERT 0 64
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 11:02:23.070821-03
(1 row)
INSERT 0 128
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 11:02:51.662401-03
(1 row)
INSERT 0 256
postgres=# select version();
                                         version                                          
------------------------------------------------------------------------------------------
 PostgreSQL 9.5.7 on x86_64-pc-linux-gnu, compiled by gcc (Debian 4.9.2-10) 4.9.2, 64-bit
(1 row)
pinogy@workstation:~ $ top -b -d 30 -p 3988 | egrep "(load|pg_cron|USED)"
top - 10:57:44 up 37 min,  6 users,  load average: 0.25, 0.20, 0.13
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
 3988 pinogy    20   0  232452   8700   6624   8700      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 10:58:14 up 38 min,  6 users,  load average: 0.29, 0.22, 0.13
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
 3988 pinogy    20   0  232452   8700   6624   8700      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 10:58:44 up 38 min,  6 users,  load average: 0.25, 0.21, 0.13
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
 3988 pinogy    20   0  232452   8700   6624   8700      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 10:59:14 up 39 min,  6 users,  load average: 0.22, 0.21, 0.13
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
 3988 pinogy    20   0  239028  11208   8896  11208      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 10:59:44 up 39 min,  6 users,  load average: 0.13, 0.18, 0.13
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
 3988 pinogy    20   0  239028  11208   8896  11208      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 11:00:14 up 40 min,  6 users,  load average: 0.08, 0.17, 0.12
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
 3988 pinogy    20   0  239028  11272   8896  11272      0 S   0.0  0.0   0:00.01 postgres: bgworker: pg_cron_scheduler
top - 11:00:44 up 40 min,  6 users,  load average: 0.05, 0.15, 0.11
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
 3988 pinogy    20   0  239028  11272   8896  11272      0 S   0.0  0.0   0:00.01 postgres: bgworker: pg_cron_scheduler
top - 11:01:14 up 41 min,  6 users,  load average: 0.03, 0.13, 0.11
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
 3988 pinogy    20   0  240120  11596   8896  11596      0 S   0.0  0.0   0:00.02 postgres: bgworker: pg_cron_scheduler
top - 11:01:44 up 41 min,  6 users,  load average: 0.02, 0.12, 0.10
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
 3988 pinogy    20   0  240120  11596   8896  11596      0 S   0.0  0.0   0:00.02 postgres: bgworker: pg_cron_scheduler
top - 11:01:45 up 41 min,  6 users,  load average: 0.02, 0.12, 0.10
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
 3988 pinogy    20   0  240120  11596   8896  11596      0 S   0.0  0.0   0:00.02 postgres: bgworker: pg_cron_scheduler
top - 11:02:15 up 42 min,  6 users,  load average: 0.01, 0.11, 0.10
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
 3988 pinogy    20   0  243640  12836   8928  12836      0 S   0.1  0.0   0:00.05 postgres: bgworker: pg_cron_scheduler
top - 11:02:45 up 42 min,  6 users,  load average: 0.00, 0.09, 0.09
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
 3988 pinogy    20   0  243640  13100   8936  13100      0 S   0.0  0.0   0:00.06 postgres: bgworker: pg_cron_scheduler
top - 11:03:15 up 43 min,  6 users,  load average: 0.00, 0.08, 0.09
```

**Comment by marcocitus** [comment]
Thanks for reporting, could you try this fix?
https://github.com/citusdata/pg_cron/tree/fix_memory_leak

**Comment by DanielCristianPinogy** [comment]
I tried the change, but the problem persists:

```
top -b -d 30 -p 10945 | egrep "(load|pg_cron|USED)"
top - 20:19:27 up  4:10,  4 users,  load average: 0.02, 0.08, 0.19
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  232452   7776   5736   7776      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 20:19:57 up  4:10,  4 users,  load average: 0.07, 0.08, 0.19
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  232452   9476   7380   9476      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 20:20:27 up  4:11,  4 users,  load average: 0.04, 0.07, 0.18
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  232452   9476   7380   9476      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 20:20:57 up  4:11,  4 users,  load average: 0.02, 0.07, 0.18
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  232452   9476   7380   9476      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 20:21:27 up  4:12,  4 users,  load average: 0.01, 0.06, 0.17
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  232452   9476   7380   9476      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 20:21:57 up  4:12,  4 users,  load average: 0.01, 0.05, 0.16
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  232452   9476   7380   9476      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 20:22:28 up  4:13,  4 users,  load average: 0.00, 0.05, 0.16
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  238844  11624   9452  11624      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 20:22:58 up  4:13,  4 users,  load average: 0.00, 0.04, 0.15
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  238828  11624   9452  11624      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 20:23:28 up  4:14,  4 users,  load average: 0.00, 0.03, 0.15
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  238836  11676   9452  11676      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 20:23:58 up  4:14,  4 users,  load average: 0.00, 0.03, 0.14
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  238836  11676   9452  11676      0 S   0.0  0.0   0:00.01 postgres: bgworker: pg_cron_scheduler
top - 20:24:28 up  4:15,  4 users,  load average: 0.00, 0.03, 0.13
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  239932  11992   9452  11992      0 S   0.0  0.0   0:00.02 postgres: bgworker: pg_cron_scheduler
top - 20:24:58 up  4:15,  4 users,  load average: 0.00, 0.02, 0.13
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  239932  11992   9452  11992      0 S   0.0  0.0   0:00.02 postgres: bgworker: pg_cron_scheduler
top - 20:25:28 up  4:16,  4 users,  load average: 0.00, 0.02, 0.12
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  241168  12688   9452  12688      0 S   0.0  0.0   0:00.03 postgres: bgworker: pg_cron_scheduler
top - 20:25:58 up  4:16,  4 users,  load average: 0.00, 0.02, 0.12
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  241168  12992   9504  12992      0 S   0.0  0.0   0:00.03 postgres: bgworker: pg_cron_scheduler
top - 20:26:28 up  4:17,  4 users,  load average: 0.00, 0.01, 0.11
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  248476  15816   9504  15816      0 S   0.1  0.0   0:00.07 postgres: bgworker: pg_cron_scheduler
top - 20:26:58 up  4:17,  4 users,  load average: 0.00, 0.01, 0.11
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  248476  16476   9652  16476      0 S   0.0  0.1   0:00.08 postgres: bgworker: pg_cron_scheduler
top - 20:27:28 up  4:18,  4 users,  load average: 0.00, 0.01, 0.10
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  277980  30244   9780  30244      0 S   0.4  0.1   0:00.20 postgres: bgworker: pg_cron_scheduler
top - 20:27:58 up  4:18,  4 users,  load average: 0.00, 0.00, 0.09
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  277980  33628  10012  33628      0 S   0.1  0.1   0:00.22 postgres: bgworker: pg_cron_scheduler
top - 20:28:28 up  4:19,  4 users,  load average: 0.00, 0.00, 0.09
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  396420 119356  10012 119356      0 S   2.5  0.4   0:00.98 postgres: bgworker: pg_cron_scheduler
top - 20:28:58 up  4:19,  4 users,  load average: 0.00, 0.00, 0.09
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  396420 121468  10012 121468      0 S   0.1  0.4   0:01.01 postgres: bgworker: pg_cron_scheduler
top - 20:29:28 up  4:20,  4 users,  load average: 0.00, 0.00, 0.08
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  452320 192124  10012 192124      0 S   3.5  0.6   0:02.06 postgres: bgworker: pg_cron_scheduler
top - 20:29:58 up  4:20,  4 users,  load average: 0.00, 0.00, 0.08
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  452320 193708  10012 193708      0 S   0.1  0.6   0:02.08 postgres: bgworker: pg_cron_scheduler
top - 20:30:28 up  4:21,  4 users,  load average: 0.00, 0.00, 0.08
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  532608 270776  10012 270776      0 S   3.2  0.8   0:03.03 postgres: bgworker: pg_cron_scheduler
top - 20:30:58 up  4:21,  4 users,  load average: 0.07, 0.02, 0.08
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
10945 pinogy    20   0  532608 273152  10012 273152      0 S   0.1  0.8   0:03.05 postgres: bgworker: pg_cron_scheduler
top - 20:31:28 up  4:22,  4 users,  load average: 0.04, 0.01, 0.08
```

Every 30 s I doubled the jobs, as seen bellow:

```
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) values ('* * * * *', 'select 1', 'localhost', '5432', 'postgres', 'pinogy');
             now              
------------------------------
 2017-06-01 20:21:30.99576-03
(1 row)

INSERT 0 1
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 20:21:59.907818-03
(1 row)

INSERT 0 1
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 20:22:31.019787-03
(1 row)

INSERT 0 2
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 20:23:01.979817-03
(1 row)

INSERT 0 4
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 20:23:29.955857-03
(1 row)

INSERT 0 8
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 20:23:58.723879-03
(1 row)

INSERT 0 16
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 20:24:29.211943-03
(1 row)

INSERT 0 32
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 20:25:00.843954-03
(1 row)

INSERT 0 64
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 20:25:29.540012-03
(1 row)

INSERT 0 128
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 20:26:00.876037-03
(1 row)

INSERT 0 256
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 20:26:31.724159-03
(1 row)

INSERT 0 512
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 20:27:03.884144-03
(1 row)

INSERT 0 1024
postgres=# select now(); insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;
              now              
-------------------------------
 2017-06-01 20:27:32.708078-03
(1 row)

INSERT 0 2048
```

One thing I noticed in this test is that it had more connections than max_connections, creating a lot of messages like these:

```
FATAL:  sorry, too many clients already
FATAL:  sorry, too many clients already
FATAL:  sorry, too many clients already
LOG:  pg_cron job 2342: connection failed
LOG:  pg_cron job 1489: connection failed
LOG:  pg_cron job 674: connection failed
```

I configured the instance to allow 10.000 connections and kept 4096 jobs, restarting PostgreSQL I got:

```
top -b -d 30 -p 29392 | egrep "(load|pg_cron|USED)"
top - 20:52:00 up  4:42,  4 users,  load average: 0.08, 0.04, 0.06
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29392 pinogy    20   0  858856  69116  11136  69116      0 S   0.0  0.2   0:00.38 postgres: bgworker: pg_cron_scheduler
top - 20:52:24 up  4:43,  4 users,  load average: 0.06, 0.03, 0.06
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29392 pinogy    20   0  858856 125084  11136 125084      0 S   2.8  0.4   0:01.04 postgres: bgworker: pg_cron_scheduler
top - 20:52:54 up  4:43,  4 users,  load average: 0.22, 0.08, 0.07
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29392 pinogy    20   0  858856 127988  11136 127988      0 S   0.1  0.4   0:01.07 postgres: bgworker: pg_cron_scheduler
top - 20:53:24 up  4:44,  4 users,  load average: 0.20, 0.08, 0.07
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29392 pinogy    20   0  947144 212608  11136 212608      0 S   3.9  0.6   0:02.23 postgres: bgworker: pg_cron_scheduler
top - 20:53:54 up  4:44,  4 users,  load average: 0.12, 0.07, 0.07
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29392 pinogy    20   0  947144 214984  11136 214984      0 S   0.1  0.7   0:02.26 postgres: bgworker: pg_cron_scheduler
top - 20:54:24 up  4:45,  4 users,  load average: 0.07, 0.07, 0.07
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29392 pinogy    20   0 1025940 297456  11136 297456      0 S   3.7  0.9   0:03.37 postgres: bgworker: pg_cron_scheduler
top - 20:54:54 up  4:45,  4 users,  load average: 0.18, 0.09, 0.08
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29392 pinogy    20   0 1025940 300096  11136 300096      0 S   0.1  0.9   0:03.39 postgres: bgworker: pg_cron_scheduler
top - 20:55:24 up  4:46,  4 users,  load average: 0.11, 0.08, 0.07
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29392 pinogy    20   0 1115876 377208  11136 377208      0 S   3.3  1.1   0:04.38 postgres: bgworker: pg_cron_scheduler
top - 20:55:55 up  4:46,  4 users,  load average: 0.06, 0.07, 0.07
```

**Comment by DanielCristianPinogy** [comment]
The values above are for PostgreSQL 9.5. I though they could be different for the current version, but the same behavior on PostgreSQL 9.6.3:

```
top - 06:29:26 up 18 min,  4 users,  load average: 0.16, 0.11, 0.07
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0  722572  12272  10164  12272      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 06:29:56 up 18 min,  4 users,  load average: 0.28, 0.15, 0.08
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0  722572  12272  10164  12272      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 06:30:26 up 19 min,  4 users,  load average: 0.17, 0.13, 0.08
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0  722572  14004  11856  14004      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 06:30:56 up 19 min,  4 users,  load average: 0.10, 0.12, 0.08
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0  722572  14004  11856  14004      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 06:31:26 up 20 min,  4 users,  load average: 0.06, 0.11, 0.07
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0  729032  16836  14620  16836      0 S   0.0  0.1   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 06:31:56 up 20 min,  4 users,  load average: 0.16, 0.13, 0.08
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0  729032  16836  14620  16836      0 S   0.0  0.1   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 06:32:26 up 21 min,  4 users,  load average: 0.09, 0.11, 0.08
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0  729136  16920  14620  16920      0 S   0.0  0.1   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 06:32:56 up 21 min,  4 users,  load average: 0.11, 0.12, 0.08
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0  729136  16920  14620  16920      0 S   0.0  0.1   0:00.01 postgres: bgworker: pg_cron_scheduler
top - 06:33:26 up 22 min,  4 users,  load average: 0.07, 0.10, 0.08
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0  730020  17276  14620  17276      0 S   0.0  0.1   0:00.01 postgres: bgworker: pg_cron_scheduler
top - 06:33:56 up 22 min,  4 users,  load average: 0.04, 0.09, 0.08
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0  730020  17276  14620  17276      0 S   0.0  0.1   0:00.01 postgres: bgworker: pg_cron_scheduler
top - 06:34:26 up 23 min,  4 users,  load average: 0.02, 0.08, 0.07
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0  733880  18808  14688  18808      0 S   0.0  0.1   0:00.02 postgres: bgworker: pg_cron_scheduler
top - 06:34:56 up 23 min,  4 users,  load average: 0.01, 0.07, 0.07
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0  733880  19124  14748  19124      0 S   0.0  0.1   0:00.03 postgres: bgworker: pg_cron_scheduler
top - 06:35:26 up 24 min,  4 users,  load average: 0.06, 0.08, 0.07
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0  747912  28468  14788  28468      0 S   0.3  0.1   0:00.12 postgres: bgworker: pg_cron_scheduler
top - 06:35:56 up 24 min,  4 users,  load average: 0.09, 0.09, 0.08
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0  747912  29316  14852  29316      0 S   0.0  0.1   0:00.13 postgres: bgworker: pg_cron_scheduler
top - 06:36:26 up 25 min,  4 users,  load average: 0.05, 0.08, 0.07
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0  817232  68436  14972  68436      0 S   1.3  0.2   0:00.52 postgres: bgworker: pg_cron_scheduler
top - 06:36:56 up 25 min,  4 users,  load average: 0.03, 0.07, 0.07
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0  817232  71816  15200  71816      0 S   0.1  0.2   0:00.56 postgres: bgworker: pg_cron_scheduler
top - 06:37:26 up 26 min,  4 users,  load average: 0.02, 0.06, 0.07
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0  930232 162748  15244 162748      0 S   4.1  0.5   0:01.80 postgres: bgworker: pg_cron_scheduler
top - 06:37:56 up 26 min,  4 users,  load average: 0.08, 0.07, 0.07
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0  930232 165124  15244 165124      0 S   0.1  0.5   0:01.84 postgres: bgworker: pg_cron_scheduler
top - 06:38:26 up 27 min,  4 users,  load average: 0.05, 0.06, 0.07
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
17017 pinogy    20   0 1012144 264812  15244 264812      0 S   4.0  0.8   0:03.04 postgres: bgworker: pg_cron_scheduler
top - 06:38:56 up 27 min,  4 users,  load average: 0.03, 0.05, 0.07
```

I did the same measurements as before:
First measurement after starting
Second after create extension pg_cron
Third after creating 1 job
Following measurements after doubling the jobs
The last 5 measurements were all taken with 4096 jobs

Command to create a job:
insert into cron.job (schedule, command, nodename, nodeport, database, username) values ('* * * * *', 'select 1', 'localhost', '5432', 'postgres', 'pinogy');

Command to double the jobs:
insert into cron.job (schedule, command, nodename, nodeport, database, username) select schedule, command, nodename, nodeport, database, username from cron.job;

**Comment by marcocitus** [comment]
Thanks for the detailed report. I updated the branch to improve cleanup code and deal with high job counts.

https://github.com/citusdata/pg_cron/tree/fix_memory_leak

It now restricts the number tasks that can run concurrently. When there are more than `cron.max_running_tasks` tasks running (default 32), it waits until one of the current tasks is done before starting a new one.

I've been testing with 10,000 jobs generated with  `SELECT cron.schedule('* * * * *', 'SELECT 1') FROM generate_series(1,10000);` and RSS is stable at around 18MB.

**Comment by DanielCristianPinogy** [comment]
Thank you for the patch, looks like the memory usage is now stable:

```
top - 14:48:18 up  1:19,  7 users,  load average: 0.93, 0.85, 0.89
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29540 pinogy    20   0 10.358g  13536  11280  13536      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 14:49:18 up  1:20,  8 users,  load average: 0.39, 0.71, 0.83
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29540 pinogy    20   0 10.364g  15472  13172  15472      0 S   0.0  0.0   0:00.00 postgres: bgworker: pg_cron_scheduler
top - 14:50:18 up  1:21,  8 users,  load average: 0.22, 0.61, 0.79
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29540 pinogy    20   0 10.364g  15492  13172  15492      0 S   0.0  0.0   0:00.01 postgres: bgworker: pg_cron_scheduler
top - 14:51:18 up  1:22,  8 users,  load average: 0.21, 0.53, 0.75
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29540 pinogy    20   0 10.364g  15492  13172  15492      0 S   0.0  0.0   0:00.01 postgres: bgworker: pg_cron_scheduler
top - 14:52:18 up  1:23,  8 users,  load average: 0.18, 0.47, 0.71
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29540 pinogy    20   0 10.364g  15492  13172  15492      0 S   0.0  0.0   0:00.01 postgres: bgworker: pg_cron_scheduler
top - 14:53:18 up  1:24,  8 users,  load average: 0.06, 0.38, 0.67
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29540 pinogy    20   0 10.364g  15492  13172  15492      0 S   0.0  0.0   0:00.03 postgres: bgworker: pg_cron_scheduler
top - 14:54:18 up  1:25,  8 users,  load average: 0.07, 0.33, 0.63
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29540 pinogy    20   0 10.364g  15708  13308  15708      0 S   0.0  0.0   0:00.03 postgres: bgworker: pg_cron_scheduler
top - 14:55:18 up  1:26,  8 users,  load average: 0.13, 0.30, 0.60
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29540 pinogy    20   0 10.364g  15708  13308  15708      0 S   0.0  0.0   0:00.04 postgres: bgworker: pg_cron_scheduler
top - 14:56:18 up  1:27,  8 users,  load average: 0.05, 0.24, 0.56
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29540 pinogy    20   0 10.364g  15708  13308  15708      0 S   0.0  0.0   0:00.04 postgres: bgworker: pg_cron_scheduler
top - 14:57:18 up  1:28,  8 users,  load average: 0.21, 0.27, 0.54
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29540 pinogy    20   0 10.364g  15708  13308  15708      0 S   0.0  0.0   0:00.05 postgres: bgworker: pg_cron_scheduler
top - 14:58:18 up  1:29,  8 users,  load average: 0.08, 0.22, 0.51
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29540 pinogy    20   0 10.364g  15708  13308  15708      0 S   0.0  0.0   0:00.06 postgres: bgworker: pg_cron_scheduler
top - 14:59:18 up  1:30,  8 users,  load average: 0.03, 0.17, 0.47
  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S  %CPU %MEM     TIME+ COMMAND
29540 pinogy    20   0 10.364g  15708  13308  15708      0 S   0.0  0.0   0:00.06 postgres: bgworker: pg_cron_scheduler
top - 15:00:18 up  1:31,  8 users,  load average: 0.08, 0.16, 0.45
```

It seems this last patch removed the bad behavior of increasingly add more memory. These are tests on my local instance, I'll apply the patch to the development server and return an OK on Monday with more information about how it behaved during the weekend.

**Comment by DanielCristianPinogy** [comment]
@marcocitus : the patch is working well.

During the weekend the process reduced the memory as seen bellow

```
top - 15:06:20 up 3 days, 19:05,  1 user,  load average: 0.00, 0.01, 0.05
Tasks:   1 total,   0 running,   1 sleeping,   0 stopped,   0 zombie
%Cpu0  :  1.6 us,  0.5 sy,  0.0 ni, 97.3 id,  0.5 wa,  0.0 hi,  0.0 si,  0.1 st
KiB Mem :  1883852 total,   866156 free,   174160 used,   843536 buff/cache
KiB Swap:  2113532 total,  2034480 free,    79052 used.  1386552 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S %CPU %MEM
15570 postgres  20   0  351888   4924   2800   4924      0 S  0.0  0.3
```

turned into after 3 days:

```
top - 15:05:46 up 6 days, 19:05,  0 users,  load average: 0.01, 0.04, 0.05
Tasks:   1 total,   0 running,   1 sleeping,   0 stopped,   0 zombie
%Cpu0  :  0.7 us,  0.4 sy,  0.0 ni, 98.9 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem :  1883852 total,    85252 free,   421948 used,  1376652 buff/cache
KiB Swap:  2113532 total,  2000104 free,   113428 used.   777052 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR   USED   SWAP S %CPU %MEM
15570 postgres  20   0  351936   3620   2788   5004   1384 S  0.0  0.2
```

Thank you very much.

**Comment by DanielCristianPinogy** [comment]
@marcocitus I noticed the behavior happening again when migrated to PostgreSQL 10. I supposed this patch was merged into master, which I'm using right now.

```
top - 23:10:30 up 43 days, 13:05,  1 user,  load average: 0.15, 0.16, 0.21
Tasks: 355 total,   1 running, 354 sleeping,   0 stopped,   0 zombie
%Cpu(s):  1.9 us,  0.3 sy,  0.0 ni, 97.8 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem : 40068444 total,   297696 free, 26094576 used, 13676172 buff/cache
KiB Swap:  4194300 total,  2134224 free,  2060076 used.  5723908 avail Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 7884 postgres  20   0 26.681g 0.021t   6216 S   0.0 55.3   4:48.03 postgres: bgworker: pg_cron scheduler
```

I can't find the branch anymore to test it against PG 10.

**Comment by DanielCristianPinogy** [comment]
Sorry, I found an old source and found that the current code has all the changes from the previous branch with a few differences, but basically the same routines with indentation and variable name changes.

I'll analyze the new scenario and create a new issue if applicable.

**Comment by hasancansaral** [comment]
I'm experiencing huge memory allocation (~550 MB per job). Would you say it's normal or am I running into an issue here?
